### PR TITLE
Fix last bit of #2330

### LINKED ISSUE #2330: Final DMC energy varies with equilibration strategy
The final DMC energy for a 20 atom cell of SrCoO3 differs when QMCPACK is run in two ways that should be equivalent:

1. A single DMC run with a timestep of 0.002/Ha.
2. DMC run first with a timestep of 0.02/Ha followed by 0.002/Ha.

In this case, a difference of 6.6(7) mHa/f.u. is observed in the final DMC energy using the CPU code on Theta (case 1) and CADES (case 1 and 2) with Theta and CADES agreeing for case 1.

A plot of the 2-timestep data with differing numbers of equilibration blocks removed, seems to show a very long and slow transient.  For comparison, a full equilibration from VMC to the DMC 0.002/Ha equilibrium occurs within 100 blocks.  Starting from the 0.02/Ha equilibrium, the new 0.002/Ha equilibrium is not reached even after 1000 blocks.

GS4_ediff

Given the rapid equlibration from the high energy VMC state, it is unclear what mechanism could be responsible for the lack of equilibration from the lower energy, and presumably closer, DMC 0.02/Ha state.  More investigation of this effect is warranted.

The code's inability to relax to the same minimum from different initial states casts doubt on the equilibrium found from standard single shot calculations performed in production.

Issue identified by @mcbennet.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
complete fix of #2330 
#2177 changed from
```
W.setActive(iat);
if (rr > m_r2max)
{
  ++nRejectTemp;
  continue;
}
W.makeMoveAndCheck(iat, dr)
```
to
```
if (rr > m_r2max)
{
  ++nRejectTemp;
  continue;
}
W.makeMoveAndCheck(iat, dr)
```
by shifting all the computation in setActive into makeMoveAndCheck.
So the early continue caused problem. The fix moves (rr > m_r2max) check after makeMoveAndCheck.
Now the failing deterministic test passes in real build.
The complex build still fails due to diverging behaviour between fix node and fix phase approximation. I have some thoughts but not fully hammered, so I will leave the resolution later.

This bug should only affect systems with very small cell (a few atoms <10) and very large time step.

## What type(s) of changes does this code introduce?
- Bugfix

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
Bora workstation.
## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp [hunk 1]
```diff
@@ -83,13 +83,11 @@ void DMCUpdatePbyPWithRejectionFast::advanceWalker(Walker_t& thisWalker, bool re
       dr += sqrttau * deltaR[iat];
       RealType rr = tauovermass * dot(deltaR[iat], deltaR[iat]);
       rr_proposed += rr;
-      if (rr > m_r2max)
+      if (!W.makeMoveAndCheck(iat, dr) || rr > m_r2max)
```
Comment (prckent) [inline src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp:86]: No function calls like this in if statements please. The execution order is not obvious. Split.

File: src/QMCDrivers/DMC/SODMCUpdatePbyPFast.cpp [hunk 2]
```diff
@@ -85,13 +85,11 @@ void SODMCUpdatePbyPWithRejectionFast::advanceWalker(Walker_t& thisWalker, bool
       ds += std::sqrt(tauovermass/spinMass)*deltaS[iat];
       RealType rr = tauovermass * dot(deltaR[iat], deltaR[iat]);
       rr_proposed += rr;
-      if (rr > m_r2max)
+      if (!W.makeMoveAndCheckWithSpin(iat, dr, ds) || rr > m_r2max)
```
Comment (prckent) [inline src/QMCDrivers/DMC/SODMCUpdatePbyPFast.cpp:88]: No function calls like this in if statements please. The execution order is not obvious. Split.

File: src/QMCDrivers/RMC/RMCUpdatePbyP.cpp [hunk 3]
```diff
@@ -153,13 +153,11 @@ void RMCUpdatePbyPWithDrift::advanceWalkersVMC()
       dr += sqrttau * deltaR[iat];
       RealType rr = tauovermass * dot(deltaR[iat], deltaR[iat]);
       rr_proposed += rr;
-      if (rr > m_r2max)
+      if (!W.makeMoveAndCheck(iat, dr) || rr > m_r2max)
```
Comment (prckent) [inline src/QMCDrivers/RMC/RMCUpdatePbyP.cpp:156]: No function calls like this in if statements please. The execution order is not obvious. Split.

File: src/QMCDrivers/RMC/RMCUpdatePbyP.cpp [hunk 4]
```diff
@@ -284,13 +282,11 @@ void RMCUpdatePbyPWithDrift::advanceWalkersRMC()
       //RealType rr=dot(dr,dr);
       RealType rr = tauovermass * dot(deltaR[iat], deltaR[iat]);
       rr_proposed += rr;
-      if (rr > m_r2max)
+      if (!W.makeMoveAndCheck(iat, dr) || rr > m_r2max)
```
Comment (prckent) [inline src/QMCDrivers/RMC/RMCUpdatePbyP.cpp:285]: No function calls like this in if statements please. The execution order is not obvious. Split.

File: src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp (hunk 1 above)
Comment (prckent) [reply src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp:86]: Rationale: many people will not know if makeMoveAndCheck is always called (or not).

Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp:86]: Actually to me
1. makeMoveAndCheck is solid.
2. rr > m_r2max is a hack. Not clear what is this actually for. This type of hack always sounds like to make bad WF caused bad moves to continue.

Comment (prckent) [reply src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp:86]: Getting rid of the maximum displacement check is a discussion for another day.

Interesting that the intent to do something smart about large moves -- which should not happen in healthy runs -- might have precipitated a far more serious bug that affected actually healthy runs.

Comment (prckent) [reply src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp:86]: Just pull the call() out of the if statements for now.

Comment (ye-luo) [reply src/QMCDrivers/DMC/DMCUpdatePbyPFast.cpp]: Done. All pulled out.

**Pull request merged**